code/datums/components/chameleon/action.dm

```diff
@@ -0,0 +1,241 @@
+#define CHAMELEON_MANUAL_COOLDOWN 2 MINUTES
+
+/// The action for the chameleon panel.
+/datum/action/chameleon_panel
+	name = "Chameleon Outfit Panel"
+	button_icon_state = "chameleon_outfit"
+	var/opened_message = FALSE
+	COOLDOWN_DECLARE(next_manual)
+
+/datum/action/chameleon_panel/Trigger()
+	if(!IsAvailable())
+		return
+	if(!opened_message)
+		to_chat(owner, "<span class='warning'>The chameleon panel UI may take a few seconds to load upon opening it for the first time! Click the action button again if it doesn't load after a few seconds!</span>")
+		opened_message = TRUE
+	ui_interact(owner)
+
+/datum/action/chameleon_panel/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ChameleonPanel")
+		ui.set_autoupdate(TRUE)
+		ui.open()
+
+/datum/action/chameleon_panel/ui_state(mob/user)
+	return GLOB.not_incapacitated_state
+
+/datum/action/chameleon_panel/ui_static_data(mob/user)
+	. = list(
+		"outfits" = list(),
+		"icons" = list(
+			"outfits" = list()
+		)
+	)
+	var/list/assets_to_send = list()
+	for(var/outfit_path in subtypesof(/datum/outfit/job))
+		var/datum/outfit/outfit = outfit_path
+		if(initial(outfit.can_be_admin_equipped))
+			.["outfits"] += list(list(
+				"name" = initial(outfit.name),
+				"type" = "[outfit]"
+			))
+			var/outfit_asset = get_outfit_icon(outfit_path)
+			if(outfit_asset)
+				.["icons"]["outfits"]["[outfit]"] = SSassets.transport.get_asset_url(outfit_asset)
+				assets_to_send += outfit_asset
+			CHECK_TICK
+	sortTim(.["outfits"], GLOBAL_PROC_REF(cmp_list_name_asc))
+	if(user.client && LAZYLEN(assets_to_send))
+		SSassets.transport.send_assets(user.client, assets_to_send)
+
+/datum/action/chameleon_panel/ui_data(mob/user)
+	var/mob/living/l_user = isliving(user) ? user : null
+	. = list(
+		"chameleon_items" = list(),
+		"manual" = list(
+			"can_craft" = HAS_TRAIT(user, TRAIT_CHAMELEON_USER) || (l_user?.mind && HAS_TRAIT(l_user.mind, TRAIT_CHAMELEON_USER)),
+			"cooldown" = round(max(COOLDOWN_TIMELEFT(src, next_manual) / 10, 0))
+		)
+	)
+	var/list/names = list()
+	for(var/C in get_chameleon_items(user))
+		var/datum/component/chameleon/chameleon = C
+		var/item_name = name_slot(user, names, chameleon)
+		var/slot_name = user.get_inventory_slot_name(chameleon.parent)
+		names += item_name
+		var/list/disguise_choices = list()
+		for(var/D in chameleon.disguise_paths)
+			var/obj/item/disguise = D
+			disguise_choices += list(list(
+				"name" = initial(disguise.name),
+				"icon_name" = initial(disguise.icon_state),
+				"type" = "[D]"
+			))
+			CHECK_TICK
+		sortTim(disguise_choices, GLOBAL_PROC_REF(cmp_list_name_asc))
```

Comment:
honestly might be best to sort by typepath of the item - it would make a lot more sense, duffel bags will show in order etc